### PR TITLE
feat(policies): add cosmos3 (Qwen3-VL-32B reasoner + flow-matching expert)

### PR DESCRIPTION
**Pull request by WilliamYue37**
## What this does

Adds a new VLA policy **`cosmos3`** (🗃️ Feature): the π0.5 flow-matching recipe on a **frozen Qwen3-VL-32B** backbone — the **reasoning tower of NVIDIA `Cosmos3-Super`**, extracted into a standalone Qwen3-VL-32B checkpoint by `src/opentau/scripts/extract_cosmos3_reasoner.py` — paired with a custom **sub-1B Qwen3-style flow-matching action expert** (~0.91B trainable).

How it works (π0.5 dual-stream, specialized for a frozen reasoner):

- **Max-batch benchmarks** (8× A100-80GB, DeepSpeed, expert-only) — see the PR comment: **ZeRO-2 → 27** micro-batch/GPU, **ZeRO-3 → ~128** (ZeRO-3 shards the replicated 64 GB backbone).
- **End-to-end ZeRO-3 max-batch run with the real extracted reasoner** (8× A100-80GB): loaded `TensorAuto/cosmos3-reason-32b` as the frozen backbone, real Qwen3-VL processor (2 cams @224 + prompt), `train_expert_only`, **B=128/GPU (1024 global)** under DeepSpeed ZeRO-3 — peak **71.1 GB / 80 GB**. On a fixed batch + fixed flow-matching noise/time (`max_delay=0`, deterministic) the loss **strictly decreased every step** (2.349 → 2.000 over 8 steps), expert trainable + backbone frozen — confirming the real reasoner trains end to end at the ZeRO-3 max batch.
- The scheduled nightly `regression_test.yml` (AWS g6) runs automatically and was not triggered manually.

## How to checkout & try? (for the reviewer)

```bash
pytest -sx tests/policies/test_cosmos3_cpu.py
```
```bash
pytest -m "gpu" tests/policies/test_cosmos3_gpu.py
```
```bash
opentau-train --config_path=configs/examples/cosmos3_training_config.json
```

## Checklist

- [x] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [x] My PR includes policy-related changes.
  - [x] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### Note: Before submitting this PR, please read the [contributor guideline](https://github.com/TensorAuto/OpenTau/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by shuheng-liu — DISMISSED** [review]

**Comment by WilliamYue37** [comment]
## Benchmark — max per-GPU batch size for training the action expert (DeepSpeed ZeRO-2 vs ZeRO-3)

**Setup.** 8× A100-SXM4-80GB (NVLink mesh). Training **only the action expert** (`train_expert_only=True`, frozen 32B Qwen3-VL backbone), via DeepSpeed. The backbone is built with **random bf16 weights from the Qwen3-VL-32B config** — GPU memory is weight-value-independent, so this equals the real Cosmos3-Super reasoner backbone — at **34.3B total / 0.91B trainable**. Input: 2 cameras @224 (prefix ≈ 151 tokens), `chunk_size=50`. Each row is a real forward + backward + optimizer step; the sweep crashes at the first OOM, so the last fitting row is the max.

### ZeRO-2 — optimizer state + gradients sharded, **parameters replicated**

| micro-batch / GPU | peak mem / GPU |
|---:|---:|
| 1 | 72.2 GB |
| 8 | 75.3 GB |
| 16 | 78.7 GB |
| 24 | 80.9 GB |
| 26 | 82.7 GB |
| **27** | **83.1 GB ← max that fits** |
| 28 | **OOM** |

Baseline after ZeRO-2 init ≈ **64 GB/GPU**: the frozen backbone is **replicated on every card**, leaving only ~15 GB of headroom. **Max micro-batch = 27.**

### ZeRO-3 — parameters **also sharded** across GPUs

| micro-batch / GPU | peak mem / GPU |
|---:|---:|
| 1 | 12.5 GB |
| 32 | 27.5 GB |
| 64 | 42.0 GB |
| 96 | 56.4 GB |
| **128** | **70.9 GB ← max (verified)** |
| 160 | **OOM** |

Baseline ≈ **8 GB/GPU** (backbone sharded 8-way); B=1 drops from 72 GB → **12.5 GB**. **Max micro-batch ≈ 128** (true ceiling ~150 by interpolation).

### Takeaways

- **Effective global batch = micro-batch × 8 GPUs × grad-accum.** At accum 1: ZeRO-2 ≈ **216/step**, ZeRO-3 ≈ **1024/step** — and arbitrarily higher with gradient accumulation (no extra memory).
- ZeRO-2's ceiling is set almost entirely by the **64 GB replicated frozen backbone**, not the trainable expert. **ZeRO-3 shards it (~8 GB/GPU) and fits ~5× the per-GPU batch** — strongly preferred when batch size matters on ≤80 GB cards.
- The numbers **scale with prefix length** (number of cameras, image resolution, prompt length); `chunk_size` has minor effect (the suffix runs through the tiny expert). For real training, leave margin below the ceiling — variable prompt/camera counts set each batch's peak.

**Repro notes.** DeepSpeed ZeRO-{2,3} with `train_expert_only`, a plain torch `AdamW` + `zero_allow_untested_optimizer` (DeepSpeed's FusedAdam JIT-compiles a CUDA op that needs ninja/nvcc). The A100 NVLink box needs no NCCL tweaks; a containerized box with broken P2P required `NCCL_P2P_DISABLE=1`.

**Review by akshay18iitg — APPROVED** [review]
Looks good
